lib/cover.nix

```diff
@@ -0,0 +1,130 @@
+{ lib, haskellLib, pkgs }:
+
+{ name
+, version
+, library
+, tests
+}:
+
+let
+  identifier = name + "-" + version;
+  toBashArray = arr: "(" + (lib.concatStringsSep " " arr) + ")";
+
+  testsAsList = lib.attrValues tests;
+  checks      = builtins.map (d: haskellLib.check d) testsAsList;
+
+  # Exclude test modules from tix file. The Main module is
+  # hard-coded here because the Main module is not listed in
+  # "$test.config.modules" (the plan.nix) but must be excluded.
+  # Note that the name of the Main module file does not matter. So
+  # a line in your cabal file such as:
+  #   main-is: Spec.hs
+  # still generates a "Main.mix" file with the contents:
+  #   Mix "Spec.hs" ...
+  # Hence we can hardcode the name "Main" here.
+  testModules = lib.foldl' (acc: test: acc ++ test.config.modules) ["Main"] testsAsList;
+
+  # Mix information HPC will need.
+  # For libraries, where we copy the mix information from differs from
+  # where HPC should look for the mix files. For tests, they are the
+  # same location.
+  mixInfo = [ { rootDir = "${library}/share/hpc/vanilla/mix";
+                searchSubDir = "${identifier}";
+              }
+            ] ++ map (drv: { rootDir = "${drv}/share/hpc/vanilla/mix"; searchSubDir = null; }) testsAsList;
+
+  mixDirs = map (info: info.rootDir) mixInfo;
+  mixSearchDirs = map (info: if info.searchSubDir == null then info.rootDir else info.rootDir + "/" + info.searchSubDir) mixInfo;
+
+in pkgs.runCommand (identifier + "-coverage-report")
+  { buildInputs = (with pkgs; [ ghc ]); }
```

Comment:
Should we use the same GHC as the project uses here? Otherwise we might a) incur a dependency on another GHC, which can be expensive, and b) the behaviour of `hpc` might be different.

Reply:
Sounds good, fixed: https://github.com/input-output-hk/haskell.nix/pull/762/files#diff-4bba04d96602cc233c6b24d553f66222R41